code/iaas/ha/src/main/java/io/cattle/platform/ha/monitor/impl/PingInstancesMonitorImpl.java

```diff
@@ -73,36 +113,173 @@ public void computeInstanceActivateReply(Event event) {
         }
     }
 
-    protected boolean handleUnreportedKnownInstances(long agentId, Map<String, String> knownInstances,
-            ReportedInstances reportedInstances, boolean checkOnly) {
-
-        if (knownInstances != null) {
-            for (Map.Entry<String, String> knownInstance : knownInstances.entrySet()) {
-                if(!reportedInstances.byUuid.contains(knownInstance.getKey())
-                        && !reportedInstances.byExternalId.contains(knownInstance.getValue())) {
-                    if (checkOnly) {
-                        return true;
-                    } else {
-                        // If this known instance was not reported, schedule a potential restart.
-                        restart(knownInstance.getKey());
-                    }
+    /*
+     * If checkOnly is true, will raise a ContainersOutOfSync exception, indicating this should be reran with checkOnly set to false.
+     */
+    void syncContainers(Map<String, Instance> knownInstances, ReportedInstances reportedInstances, long agentAccountId, long agentId, long hostId,
```

Comment:
Update - I guess we should be fine if the syncContainers is called from within a single threaded task (PingInstanceMonitorImpl for example), so just wanted to make sure that thats the case.